File: src/services/raio-x.service.ts
import type { CompradorConhecidoDTO } from '../repositories/interfaces/index.js'

export interface RaioXLocalResult {
  retrato: { frase: string; complemento: string | null }
  fatores: Array<{ atributo: string; pesoPercentual: number; descricao: string }>
  estatisticas: {
    totalClientes: number
    ativos: number
    comRecompra: number
    percentualFieis: number
    ticketMedio?: number
    ticketMin?: number
    ticketMax?: number
  }
  segmentos: Array<{ segmento: string; quantidade: number; percentual: number }>
  potencial: { mensagem: string; cta: string }
}

const MIN_COMPRADORES = 3

export function calcularRaioXLocal(
  compradores: CompradorConhecidoDTO[],
): RaioXLocalResult {
  if (compradores.length < MIN_COMPRADORES) {
    throw Object.assign(
      new Error(
        `Mínimo de ${MIN_COMPRADORES} compradores necessários para gerar o Raio-X.`,
      ),
      { statusCode: 422 },
    )
  }

  const total = compradores.length
  const ativos = compradores.filter((c) => c.ativo).length
  const comRecompra = compradores.filter((c) => c.frequencia >= 2).length
  const percentualFieis = Math.round((comRecompra / total) * 100)

  const tickets = compradores
    .filter((c) => c.ticketMedio > 0)
    .map((c) => c.ticketMedio)
  const ticketMedio = tickets.length
    ? Math.round(tickets.reduce((a, b) => a + b, 0) / tickets.length)
    : undefined
  const ticketMin = tickets.length ? Math.min(...tickets) : undefined
  const ticketMax = tickets.length ? Math.max(...tickets) : undefined

  // Frequência de cada atributo presente em atributosOriginais
  const atributoFreq = new Map<string, number>()
  for (const c of compradores) {
    for (const [k, v] of Object.entries(c.atributosOriginais)) {
      if (v !== null && v !== undefined && String(v).trim() !== '') {
        atributoFreq.set(k, (atributoFreq.get(k) ?? 0) + 1)
      }
    }
  }

  const topAtributos = [...atributoFreq.entries()]
    .sort((a, b) => b[1] - a[1])
    .slice(0, 5)

  const totalPeso = topAtributos.reduce((sum, [, freq]) => sum + freq, 0) || 1
  const fatores = topAtributos.map(([atributo, freq]) => ({
    atributo,
    pesoPercentual: Math.round((freq / totalPeso) * 100),
    descricao: `Presente em ${Math.round((freq / total) * 100)}% dos bons clientes`,
  }))

  // Melhor atributo para segmentação: presença alta, cardinalidade moderada (2–8 valores únicos)
  const melhorAtributo = escolherAtributoDeSegmentacao(compradores, topAtributos)
  const segmentos = segmentarPor(compradores, melhorAtributo, total)

  return {
    retrato: {
      frase: composeRetrato(ativos, total, ticketMedio, percentualFieis),
      complemento: melhorAtributo
        ? `O atributo predominante é "${melhorAtributo}", presente na maioria dos bons clientes.`
        : null,
    },
    fatores,
    estatisticas: {
      totalClientes: total,
      ativos,
      comRecompra,
      percentualFieis,
      ...(ticketMedio !== undefined ? { ticketMedio } : {}),
      ...(ticketMin !== undefined ? { ticketMin } : {}),
      ...(ticketMax !== undefined ? { ticketMax } : {}),
    },
    segmentos,
    potencial: calcularPotencial(percentualFieis, ativos, total),
  }
}

// ─── helpers ─────────────────────────────────────────────────────────────────

function escolherAtributoDeSegmentacao(
  compradores: CompradorConhecidoDTO[],
  topAtributos: Array<[string, number]>,
): string | null {
  for (const [atributo] of topAtributos) {
    const valores = new Set(
      compradores
        .map((c) => c.atributosOriginais[atributo])
        .filter((v) => v !== null && v !== undefined && String(v).trim() !== ''),
    )
    if (valores.size >= 2 && valores.size <= 8) return atributo
  }
  return topAtributos[0]?.[0] ?? null
}

function segmentarPor(
  compradores: CompradorConhecidoDTO[],
  atributo: string | null,
  total: number,
): Array<{ segmento: string; quantidade: number; percentual: number }> {
  if (!atributo) return []

  const contagem = new Map<string, number>()
  for (const c of compradores) {
    const valor = String(c.atributosOriginais[atributo] ?? 'Não informado').trim()
    contagem.set(valor, (contagem.get(valor) ?? 0) + 1)
  }

  return [...contagem.entries()]
    .sort((a, b) => b[1] - a[1])
    .slice(0, 6)
    .map(([segmento, quantidade]) => ({
      segmento,
      quantidade,
      percentual: Math.round((quantidade / total) * 100),
    }))
}

function composeRetrato(
  ativos: number,
  total: number,
  ticketMedio: number | undefined,
  percentualFieis: number,
): string {
  const partes: string[] = []

  const pctAtivos = Math.round((ativos / total) * 100)
  partes.push(
    `${total} cliente${total !== 1 ? 's' : ''} analisado${total !== 1 ? 's' : ''}, ${pctAtivos}% ativo${pctAtivos !== 1 ? 's' : ''}.`,
  )

  if (ticketMedio !== undefined) {
    partes.push(
      `Ticket médio de R$ ${ticketMedio.toLocaleString('pt-BR')}.`,
    )
  }

  if (percentualFieis >= 60) {
    partes.push(`Alta fidelização: ${percentualFieis}% com recompra.`)
  } else if (percentualFieis >= 30) {
    partes.push(`Fidelização moderada: ${percentualFieis}% com recompra.`)
  } else {
    partes.push(`Fidelização baixa: ${percentualFieis}% com recompra.`)
  }

  return partes.join(' ')
}

function calcularPotencial(
  percentualFieis: number,
  ativos: number,
  total: number,
): { mensagem: string; cta: string } {
  const pctAtivos = Math.round((ativos / total) * 100)

  if (percentualFieis >= 60 && pctAtivos >= 70) {
    return {
      mensagem: 'Base saudável com alta fidelização. Momento ideal para expansão por lookalike.',
      cta: 'Execute uma análise de mercado para encontrar clientes com perfil similar.',
    }
  }

  if (percentualFieis >= 30) {
    return {
      mensagem: 'Fidelização moderada. Há espaço para melhorar retenção antes de expandir.',
      cta: 'Identifique os ativos com recompra e priorize perfis similares a eles.',
    }
  }

  return {
    mensagem: 'Fidelização abaixo do esperado. Revisar critérios de qualificação pode aumentar a conversão.',
    cta: 'Analise os atributos dos clientes que mais repetiram compra e refine o perfil ideal.',
  }
}

File: app/services/raiox_service.py
"""Raio-X service for customer ideal profile diagnostics."""

from collections import Counter
from statistics import mean

from app.schemas.contracts import (
    EstatisticasRaioX,
    FatorRaioX,
    PotencialRaioX,
    RaioXRequest,
    RaioXResponse,
    RetratoRaioX,
    SegmentoRaioX,
)


class RaioXService:
    """Builds a concise diagnostic report from known buyers."""

    def gerar(self, req: RaioXRequest) -> RaioXResponse:
        ativos = [comprador for comprador in req.compradores if comprador.ativo]
        if len(ativos) < 3:
            raise ValueError(
                "Mínimo de 3 compradores ativos para gerar o Raio-X. "
                f"Encontrados: {len(ativos)}."
            )

        compradores_fieis = [
            comprador for comprador in ativos if comprador.frequencia >= 2
        ]
        estatisticas = self._estatisticas(ativos, compradores_fieis, req)
        segmentos = self._segmentos(ativos)
        fatores = self._fatores(ativos)
        retrato = self._retrato(segmentos, estatisticas)
        potencial = self._potencial(segmentos, estatisticas)

        return RaioXResponse(
            retrato=retrato,
            fatores=fatores,
            estatisticas=estatisticas,
            segmentos=segmentos,
            potencial=potencial,
        )

    def _estatisticas(
        self,
        ativos,
        compradores_fieis,
        req: RaioXRequest,
    ) -> EstatisticasRaioX:
        tickets = [
            float(comprador.ticket_medio)
            for comprador in ativos
            if comprador.ticket_medio > 0
        ]
        percentual_fieis = (
            round(len(compradores_fieis) / len(ativos) * 100) if ativos else 0
        )

        return EstatisticasRaioX(
            total_clientes=len(req.compradores),
            ativos=len(ativos),
            com_recompra=len(compradores_fieis),
            percentual_fieis=percentual_fieis,
            ticket_medio=round(mean(tickets), 2) if tickets else None,
            ticket_min=min(tickets) if tickets else None,
            ticket_max=max(tickets) if tickets else None,
        )

    def _segmentos(self, ativos) -> list[SegmentoRaioX]:
        contador = Counter(self._segmento_do_comprador(comprador) for comprador in ativos)
        total = len(ativos)
        return [
            SegmentoRaioX(
                segmento=segmento,
                quantidade=quantidade,
                percentual=round(quantidade / total * 100),
            )
            for segmento, quantidade in contador.most_common(5)
        ]

    def _fatores(self, ativos) -> list[FatorRaioX]:
        total = len(ativos)
        fatores: list[FatorRaioX] = []

        for chave in ["cidade", "porte", "segmento", "cnae", "uf"]:
            valores = [
                self._normalizar_valor(comprador.atributos_originais.get(chave))
                for comprador in ativos
            ]
            valores = [valor for valor in valores if valor]
            if not valores:
                continue

            dominante, quantidade = Counter(valores).most_common(1)[0]
            percentual = round(quantidade / total * 100)
            fatores.append(
                FatorRaioX(
                    atributo=self._rotulo_atributo(chave),
                    peso_percentual=percentual,
                    descricao=(
                        f"{percentual}% dos clientes ativos compartilham "
                        f"{self._rotulo_atributo(chave).lower()} = {dominante}."
                    ),
                )
            )

        if not fatores:
            ticket_medio = round(
                mean(float(comprador.ticket_medio) for comprador in ativos), 2
            )
            fatores.append(
                FatorRaioX(
                    atributo="Ticket medio",
                    peso_percentual=100,
                    descricao=(
                        "A base ativa se concentra em torno de um ticket medio "
                        f"de {ticket_medio:.2f}."
                    ),
                )
            )

        fatores.sort(key=lambda fator: (-fator.peso_percentual, fator.atributo))
        return fatores[:4]

    def _retrato(
        self,
        segmentos: list[SegmentoRaioX],
        estatisticas: EstatisticasRaioX,
    ) -> RetratoRaioX:
        segmento_principal = segmentos[0].segmento if segmentos else "uma base diversificada"
        frase = f"O cliente ideal se concentra em {segmento_principal}."
        complemento = (
            f"{estatisticas.percentual_fieis}% da base ativa já apresenta recompra."
            if estatisticas.ativos > 0
            else None
        )
        return RetratoRaioX(frase=frase, complemento=complemento)

    def _potencial(
        self,
        segmentos: list[SegmentoRaioX],
        estatisticas: EstatisticasRaioX,
    ) -> PotencialRaioX:
        segmento_principal = segmentos[0].segmento if segmentos else "segmentos aderentes"
        mensagem = (
            f"A maior tração está em {segmento_principal}, com "
            f"{estatisticas.percentual_fieis}% de fidelização na base ativa."
        )
        cta = "Priorize prospecção em segmentos com o mesmo perfil dominante."
        return PotencialRaioX(mensagem=mensagem, cta=cta)

    def _segmento_do_comprador(self, comprador) -> str:
        atributos = comprador.atributos_originais
        for chave in ("segmento", "cnae", "porte", "cidade", "uf"):
            valor = self._normalizar_valor(atributos.get(chave))
            if valor:
                return f"{self._rotulo_atributo(chave)}: {valor}"
        return f"Tipo: {comprador.tipo}"

    def _rotulo_atributo(self, chave: str) -> str:
        rotulos = {
            "cidade": "Cidade",
            "porte": "Porte",
            "segmento": "Segmento",
            "cnae": "CNAE",
            "uf": "UF",
        }
        return rotulos.get(chave, chave.replace("_", " ").title())

    def _normalizar_valor(self, valor) -> str:
        if valor is None:
            return ""
        if isinstance(valor, (list, tuple, set)):
            return ", ".join(str(item).strip() for item in valor if str(item).strip())
        texto = str(valor).strip()
        return texto
